### ISSUE TITLE
"Types .. have no overlap" (2367) error when LHS and RHS generics both extend same constraint

### ISSUE THREAD

**Issue by yseymour** [microsoft/TypeScript#51584]
# Bug Report

### 🔎 Search Terms

2367 types overlap

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.0-dev.20221117#code/JYOwLgpgTgZghgYwgAgEJwM4oN7IB4BcyIArgLYBG0yAvgFB0IA2mGyAwgDwAqyEekEABM26LAD5k2OslnIEAexAYwUEgjAKoACigQ4QpUwCe+ItwCUU+jLkwFCzgFU+AiMNGYI47caJOraTlg5GAYZF9kAF4Y5DAAC2AMADo8QJpZAHpM5E4AWjzkAHInIuQ4YWLuMvi4ADcUEAVkBQaoFgAHABpkChIwOPiIU0USJiFbWXoaIA)

### 💻 Code

```ts
interface Base { x: number }

class C<T extends Base> {
    constructor(readonly x: T) {}

    foo<U extends Base>(y: U) {
        if (y === this.x) {}  // <-- error: 'U' and 'T' have no overlap
    }
}
```

### 🙁 Actual behavior

An error is raised on the `if` clause.
```
This comparison appears to be unintentional because the types 'U' and 'T' have no overlap.(2367)
```

### 🙂 Expected behavior

Both `U` and `T` are constrained by `Base`, so they conceivably _could_ overlap, so comparing them should be OK (and as a side-effect, it'd be nice if the comparison also refined `U` to `T` inside the `if` block).

maybe related: #50893

**Comment by RyanCavanaugh** [comment]
> so they conceivably could overlap

This is always going to be true, since all types overlap at `unknown`. Equality checks are necessarily incomplete and while this one is on the border, I think being a bit safer is the preferred direction to go.

**Comment by fatcerberus** [comment]
Hmm, this seems like a duplicate of #50893 which is tagged as a bug.  @RyanCavanaugh, thoughts?

**Comment by RyanCavanaugh** [comment]
The difference is between `foo<U extends Base>` vs `foo<U extends T>` (in the context of this example)

**Comment by yseymour** [comment]
> all types overlap at `unknown`.

Agreed, but that's materially different to this case where we have two types that are related by a common constraint, isn't it? 

I guess I can't see a reason for this error in cases where the common supertype is known to be neither `any` nor `unknown`.

Let me give you an example that's closer to my actual code:
```ts
enum Model { User, Group }

interface User { name: string, group: Group }
interface Group { name: string }

type RecordType = {
    [Model.User]: User;
    [Model.Group]: Group;
}

class View<T extends Model> {
    constructor(
        readonly primaryModel: T
    ) {}

    consume<U extends Model>(model: U, record: RecordType[U]) {
        if (model === this.primaryModel) {  // <-- error
            this.consumePrimaryModel(record);
        }
        else {
            this.consumeForeignModel(record);
        }
    }

    consumePrimaryModel(record: RecordType[T]) {}
    consumeForeignModel(record: RecordType[Model]) {}
}
```
([Playground](https://www.typescriptlang.org/play?ts=5.0.0-dev.20221118#code/KYOwrgtgBAsg9gE2AGygbygVQM7AE4A0UA4nnGAA5QC+AULQJYgAu+AZgIYDGwWue6KCA4RgALijZmeJgHMisspQmlyVOk1Z5OPEkqoZhoiVJkhZNeswCeFXgCVgXOHgQAVW7wC86WlH9QANrwSMgAdDj4ALoSkXgA3H4BwYgoYaqUMXpqiXS0XMgc2NhQAGoMwADuADxuUMAAHqwgCCUhKAB8vgFQziCmYFzMLgAUST3+eMAcCHAgyNZQFDIQHHjW7cgSbuNQAJToeT192JDA1Zj1TaCtsKnIHSMQ97FEU86uEo4f7p6BmFEDmhdj0GGwoE97lAvDCoMwABYMbBhZYMVbrTZA-wAemxUGqAFoCfU8GQ8CCJv4EUiwiczgAFFZrDb3EbvFwIPaJSn+Og8+rIXDdfkBanIumiABiLmADFkIE2bKcHK5FN5uyOAQlwEZaOZivZnyg3w5HjsgTcgMOu210qmcoVrMNCC+ytcZuAKVCVrQdGoQA))

A `View` can be instantiated for different kinds of `Model`, and needs to be able to consume records of either its primary model, or a foreign dependency. I want to dispatch based on the kind of model that was received. As written, the code feels correct, but doesn't type-check without additional assertions, which feels clunky (and IMO less safe, since it's possible to get the type assertion wrong).

**Comment by fatcerberus** [comment]
> that's materially different to this case where we have two types that are related by a common constraint, isn't it?

Not really.  Even "unconstrained" type parameters are implicitly constrained by `unknown`, and in that case it's usually clear that comparing them is a bug (if you wanted the types to be correlated you would've either derived them from a single type parameter or else have the second be constrained by the first).  So clearly "having a common constraint" can't be a sufficient condition for comparability and it then becomes a tricky subjective question of where to draw the line.
